src/forcing/ForcingsEngineLumpedDataProvider.cpp

```diff
@@ -0,0 +1,250 @@
+    for (std::size_t vi = 0; vi < outputs.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(outputs[vi])),
+            size
+        };
+
+        for (std::size_t di = 0; di < size; di++) {
+            // Move 0 time (current) to 1 (previous), and set the
+            // current values from var_span
+            var_cache_.at({{1, di, vi}}) = var_cache_.at({{0, di, vi}});
+            var_cache_.at({{0, di, vi}}) = var_span[di];
+        }
+    }
+}
+
+double Provider::at(
+    const std::string& divide_id,
+    const std::string& variable,
+    bool               previous
+)
+{
+    const auto d_idx = divide_index(divide_id);
+    const auto v_idx = variable_index(variable);
+
+    if (d_idx == bad_index || v_idx == bad_index) {
+        const auto shape = var_cache_.shape();
+        const auto shape_str = std::to_string(shape[0]) + ", " +
+                               std::to_string(shape[1]) + ", " +
+                               std::to_string(shape[2]);
+
+        const auto index_str = (d_idx == bad_index ? "?" : std::to_string(d_idx)) + ", " +
+                               (v_idx == bad_index ? "?" : std::to_string(v_idx));
+
+        throw std::out_of_range{
+            "Failed to get ForcingsEngineLumpedDataProvider value at index {" + index_str + "}" +
+            "\n  Shape    : {" + shape_str + "}" +
+            "\n  Divide   : `" + divide_id + "`" +
+            "\n  Variable : `" + variable  + "`"
+        };
+    }
+
+    return at(d_idx, v_idx, previous);
+}
+
+double Provider::at(
+    std::size_t divide_idx,
+    std::size_t variable_idx,
+    bool      previous
+)
+{
+    if (divide_idx == bad_index || variable_idx == bad_index) {
+        const auto shape = var_cache_.shape();
+        const auto shape_str = std::to_string(shape[0]) + ", " +
+                               std::to_string(shape[1]) + ", " +
+                               std::to_string(shape[2]);
+
+        const auto index_str = (divide_idx == bad_index ? "?" : std::to_string(divide_idx)) + ", " +
+                               (variable_idx == bad_index ? "?" : std::to_string(variable_idx));
+
+        throw std::out_of_range{
+            "Failed to get ForcingsEngineLumpedDataProvider value at index {" + index_str + "}" +
+            "\n  Shape          : {" + shape_str + "}"
+        };
+    }
+    
+    return var_cache_.at({{previous ? 1U : 0U, divide_idx, variable_idx}});
+}
+
+double Provider::get_value(
+    const CatchmentAggrDataSelector& selector,
+    ReSampleMethod m
+)
+{
+    const auto start = clock_type::from_time_t(selector.get_init_time());
+    const auto end = std::chrono::seconds{selector.get_duration_secs()} + start;
+    const auto step = std::chrono::seconds{this->record_duration()};
+    const std::string id = selector.get_id();
+    const std::string var = selector.get_variable_name();
```

Comment:
Per Slack conversation, these should go through `divide_index()` and `variable_index()` outside the loop